CMakeLists.txt

```diff
@@ -1,49 +1,63 @@
-cmake_minimum_required(VERSION 3.4)
+cmake_minimum_required(VERSION 3.4..3.18)
```

Comment:
The `<policy_max>` is also new to me, but the way I read [the docs](https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html), I think it can be very useful and I think we should specify a maximum version.

First, while it was indeed introduced in 3.12, older versions of cmake should still fallback gracefully to the behaviour of just specifying the minimum.

> If the running version of CMake is older than 3.12, the extra `...` dots will be seen as version component separators, resulting in the `...<max>` part being ignored and preserving the pre-3.12 behavior of basing policies on `<min>`.

Second, this does not specify a maximum version of cmake that can be used, but rather it  implicitly invokes `cmake_policy(VERSION <min>...<max>)`.

> All policies known to the running version of CMake and introduced in the <min> (or <max>, if specified) version or earlier will be set to use NEW behavior. All policies introduced in later versions will be unset. This effectively requests behavior preferred as of a given CMake version and tells newer CMake versions to warn about their new policies.

With this, while the minimum version that is usable remains the same (namely, 3.4), if you do use a newer version, all policies introduced in versions up to 3.18 will be set to NEW. Without this change, even for newer versions of cmake, only policies introduced in 3.4 or earlier versions will be set to NEW and all more recent policies are set to OLD (and give a warning). Remember that policies in cmake are behaviour changes, that are potentially breaking for specific existing projects. They default to the OLD behaviour (and possibly cause a warning). Basically, we are asserting that the NEW behaviour of policies up to and including 3.18 does not break our project.

Why is this useful? Some of the [policies](https://cmake.org/cmake/help/latest/manual/cmake-policies.7.html) are quite convenient for users. For example [CMP0074](https://cmake.org/cmake/help/latest/policy/CMP0074.html) introduced in 3.12 allows specifying per-package ROOT directories for `find_package` via environment (or passed cmake) variables. With this change, users of cmake 3.12 or later can use this feature when building Sophus. Without this change, they cannot (they get a warning if they try).

Reply:
I disagree. The behavior must be consistent under every supported CMake version. Therefore, the policies should be set explicitly.

Reply:
> I disagree. The behavior must be consistent under every supported CMake version. Therefore, the policies should be set explicitly.

Sorry, but this doesn't make any sense to me. Obviously, you cannot use a policy introduced in 3.12 when you use cmake 3.4., so you cannot expect "consistent" behaviour unless you set every policy later than 3.4 to OLD. But why should we stop the user of 3.12 to take advantage of the new behaviour of newer policies?

The reason the policies default to OLD is that they **potentially** break a project, not that they break **every** project. For a specific project, this can be checked. For example, since AFAICS Sophus doesn't define any `..._ROOT` variables itself, it's safe to set `CMP0074` to `NEW` (and yes, only users with cmake >= 3.12 will be able to use the feature, but it doesn't break the project for users with cmaker < 3.12).

Reply:
(But you are right that one should check the policies to see if there is one that does conflict with Sophus' cmake file. From a quick look through most of them don't matter at all here (no benefit, no harm), but that was only a very quick look.)

Reply:
Your understanding of policies seems to be flawed if my objection does not make sense to you. Obviously, policies existed before CMake 3.12 which will not be activated if you use the new range specifier.

Reply:
Thanks for the complement and constructive arguments.

If you put `cmake_minimum_required(VERSION 3.4..3.18)` and run cmake version X, then
 - a) for X > 3.18, all policies defined in version Y for Y <= 3.18 are set to NEW, and policies defined in version Y for 3.18 < Y <= X are set to OLD.
 - b) for 3.12 <= X <= 3.18, all policies known to version X are set to NEW
 - c) for 3.4 <= X < 3.12, all policies defined in version Y for Y <= 3.4 are set to NEW, and polices defined in version Y for 3.4 < Y <= X are set to OLD.

So ok, using an old version of cmake (<3.12) --- i.e. case c --- has an inconsistency w.r.t. to the policies introduced in 3.5 up to 3.11. There are 7 such policies and to me it looks like none of them are relevant for Sophus' cmake code, but if you think there is a benefit, one could explicitly set those policies to NEW, then they are also NEW in case c).

Regarding explicitly enabling NEW vs explicitly enabling OLD, to me it seems the [cmake docs](https://cmake.org/cmake/help/latest/manual/cmake-policies.7.html) are quite explicit. 

> A policy is a deprecation mechanism and not a reliable feature toggle. A policy should almost never be set to OLD, except to silence warnings in an otherwise frozen or stable codebase, or temporarily as part of a larger migration path. The OLD behavior of each policy is undesirable and will be replaced with an error condition in a future release.

According to that, one should aim to make one's code working with the NEW behaviour and switch on OLD only if it cannot be avoided. Therefore, to me it seems like `cmake_minimum_required(VERSION 3.4..3.18)` (or even `cmake_minimum_required(VERSION 3.4..3.22)`) is the way to go.

But hey, I'll not continue arguing after this, I think I've made my points. If others disagree, so be it.

Reply:
Listen, I'm not trying to argue with you. It's already hard enough to read your wall of text as is. I'm just pointing out that what you propose can be solved without fancy new CMake features in a uniform way. You are also contradicting yourself by claimining that many policies are not relevant while still willing to enable all of them. Not much of a point there, is it?

Reply:
@sergiud Let's keep the discussion polite and constructive. CMake `3.18` was the max version I've tested it with, so I've added it to be on the safe side. But I can understand that in this particular case nothing will probably break with the newer versions, so I'm fine removing it from this PR. I think we should close this topic here and if necessary create a new PR with changes at some point.